sns_ik_lib/include/sns_ik/sns_vel_ik_base.hpp

```diff
@@ -0,0 +1,183 @@
+/** @file sns_vel_ik_base.hpp
+ *
+ * @brief The file provides the basic implementation of the SNS-IK velocity solver
+ *
+ * @author Matthew Kelly
+ *
+ * This file provides a set of functions that return simple kinematic chains that are used for the
+ * unit tests. This allows unit tests to run quickly without depending on external URDF files.
+ *
+ *    Copyright 2018 Rethink Robotics
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#ifndef SNS_IK_LIB__SNS_VEL_IK_H_
+#define SNS_IK_LIB__SNS_VEL_IK_H_
+
+#include <Eigen/Dense>
+#include <memory>
+
+#include "sns_linear_solver.hpp"
+
+namespace sns_ik {
+
+enum class SnsIkExitCode {
+  Success,  // successfully solver the optimization problem
+  BadUserInput,  // user input failed basic checks (eg. inconsistent matrix size)
+  InfeasibleTask,  // there is no feasible solution to the primary task
+  InternalError  // there was an internal error in the solver (should never happen...)
+};
+
+class SnsVelIkBase {
+
+public:
+
+  // Smart pointer typedefs. Note: all derived classes MUST override these smart pointers.
+  typedef std::shared_ptr<SnsVelIkBase> Ptr;
+  typedef std::unique_ptr<SnsVelIkBase> uPtr;
+
+  /**
+   * Create a default solver with nJnt joints and no bounds on joint velocity
+   * @param nJnt: number of joints in the robot model (columns in the jacobian)
+   * @return: velocity solver iff successful, nullptr otherwise
+   */
+  static std::unique_ptr<SnsVelIkBase> create(int nJnt);
+
+  /**
+   * Create a default solver with nJnt joints and infinite bounds on the joint velocity.
+   * @param dqLow: lower bound on the velocity of each joint
+   * @param dqUpp: upper bound on the velocity of each joint
+   * @return: velocity solver iff successful, nullptr otherwise
+   */
+  static std::unique_ptr<SnsVelIkBase> create(const Eigen::ArrayXd& dqLow,
```

Comment:
Is 'nJnt' implicitly obtained from the size of the array?

Reply:
Yes.
The documentation for this method was done poorly. I fixed it.